File: C++/spiral-matrix-ii.cpp
class Solution
{
public:
    vector<vector<int>> generateMatrix(int n)
    {
        vector<vector<int>> mat(n, vector<int>(n, 0));

        int n2 = n * n;

        int left = 0;
        int right = n - 1;
        int top = 0;
        int bottom = n - 1;

        int id = 1;
        while (id <= n2)
        {
            for (int i = left; id <= n2 && i <= right; i++)
            {
                mat[top][i] = id++;
            }
            top++;
            for (int i = top; id <= n2 && i <= bottom; i++)
            {
                mat[i][right] = id++;
            }
            right--;
            for (int i = right; id <= n2 && i >= left; i--)
            {
                mat[bottom][i] = id++;
            }
            bottom--;
            for (int i = bottom; id <= n2 && i >= top; i--)
            {
                mat[i][left] = id++;
            }
            left++;
        }

        return mat;
    }
};

File: JavaScript/spiral-matrix-ii.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    const ans = new Array(n)
    const lastNum = n * n

    for (let i = 0; i < n; i++) {
        ans[i] = new Array(n).fill(0)
    }
    
    // left, down, right, up
    const dirVectors = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    let x = 1, d = 0, i = 0, j = 0

    while (x <= lastNum) {
        // check collision with borders or other numbers
        if ( i < 0 || j < 0 || i >= n || j >= n || ans[i][j] != 0) {
            // go back
            i -= dirVectors[d][0]
            j -= dirVectors[d][1]

            // turn to the next direction
            d = (d + 1) % 4

            // advance in new direction
            i += dirVectors[d][0]
            j += dirVectors[d][1]
        }

        ans[i][j] = x++
        i += dirVectors[d][0]
        j += dirVectors[d][1]
    }

    return ans
};